libraries/protocol/include/graphene/protocol/restriction.hpp

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2018 Abit More, and contributors.
+ *
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#pragma once
+#include <graphene/protocol/base.hpp>
+#include <graphene/protocol/types.hpp>
+
+namespace graphene { namespace protocol {
+
+/**
+  * Defines the set of valid operation restritions as a discriminated union type.
+  */
+struct restriction {
+   enum function_type {
+      func_eq,
+      func_ne,
+      func_lt,
+      func_le,
+      func_gt,
+      func_ge,
+      func_in,
+      func_not_in,
+      func_has_all,
+      func_has_none,
+      func_attr,
+      func_logical_or,
+      func_variant_assert,
+      FUNCTION_TYPE_COUNT ///< Sentry value which contains the number of different types
+   };
+
+   // A variant assertion argument is a pair of the tag expected to be in the variant, and the restrictions to apply
+   // to the value
+   using variant_assert_argument_type = pair<int64_t, vector<restriction>>;
+
+#define GRAPHENE_OP_RESTRICTION_ARGUMENTS_VARIADIC \
+   /*  0 */ void_t, \
+   /*  1 */ bool, \
+   /*  2 */ int64_t, \
+   /*  3 */ string, \
+   /*  4 */ time_point_sec, \
+   /*  5 */ public_key_type, \
+   /*  6 */ fc::sha256, \
```

Comment:
We may need to support `fc::ripemd160`, `fc::sha160`, `fc::sha1` and etc for `assert_operation` and HTLC operations.

Reply:
See 7c66d15